### PR TITLE
[SPARK-32193][SQL][DOCS] Update regexp usage in SQL docs

### PR DESCRIPTION
**Pull request by GuoPhilipse**
### What changes were proposed in this pull request?
update REGEXP usage and examples in sql-ref-syntx-qry-select-like.cmd

### Why are the changes needed?
make the usage of REGEXP known to more users

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

No tests

### CODE REVIEW

**Comment by maropu** [comment]
I think you don't need to file jira for this kind of minor doc fixes. Btw, any other systems supporting `REGEXP` for regular expressions other than Hive? I think we might be able to assign an alias name to it if it is a general name.

File: docs/sql-migration-guide.md [hunk 1]
```diff
@@ -970,3 +970,4 @@ Below are the scenarios in which Hive and Spark generate different results:
 * `ACOS(n)` If n < -1 or n > 1, Hive returns null, Spark SQL returns NaN.
 * `ASIN(n)` If n < -1 or n > 1, Hive returns null, Spark SQL returns NaN.
 * `CAST(n AS TIMESTAMP)` If n is integral numbers, Hive treats n as milliseconds, Spark SQL treats n as seconds.
+* `REGEXP(str, patten)` Hive support this function, Spark SQL use RLIKE instead.
```
Comment (maropu) [inline docs/sql-migration-guide.md:973]: nit: `support` -> `supports` and `use` -> `uses`

**Comment by maropu** [comment]
cc: @HyukjinKwon

**Comment by GuoPhilipse** [comment]
> I think you don't need to file jira for this kind of minor doc fixes. Btw, any other systems supporting `REGEXP` for regular expressions other than Hive? I think we might be able to assign an alias name to it if it is a general name.

So far as i know ,mysql and hive support this key word .But i think a alias name seems a good idea.
examples:
`SELECT 'abc' REGEXP '([a-z]+)';`
Ressult:
`1`

**Comment by HyukjinKwon** [comment]
Thanks @maropu. @GuoPhilipse, does MySQL also supports `REGEXP` as a function?

From reading the doc in Hive at https://cwiki.apache.org/confluence/display/Hive/LanguageManual+UDF, it's same as Rlike which Spark supports. Maybe we should alias it.

FWIW, there are many unsupported expressions explicitly:
https://github.com/apache/spark/blob/5d870ef0bc70527fd1bc99a4ad17e4941c923351/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala#L189-L192

File: docs/sql-migration-guide.md (hunk 1 above)
Comment (HyukjinKwon) [inline docs/sql-migration-guide.md:973]: Also, technically this isn's in "the scenarios in which Hive and Spark generate different results".

**Comment by GuoPhilipse** [comment]
> Thanks @maropu. @GuoPhilipse, does MySQL also supports `REGEXP` as a function?
> 
> From reading the doc in Hive at https://cwiki.apache.org/confluence/display/Hive/LanguageManual+UDF, it's same as Rlike which Spark supports. Maybe we should alias it.
> 
> FWIW, there are many unsupported expressions explicitly:
> https://github.com/apache/spark/blob/5d870ef0bc70527fd1bc99a4ad17e4941c923351/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala#L189-L192

yes, i posted the REGEXP  usage  in mysql last comment.

**Comment by HyukjinKwon** [comment]
No, @GuoPhilipse, I meant if it's supported in `SELECT REGEXP('abc', '([a-z]+)');` way.

**Comment by GuoPhilipse** [comment]
> No, @GuoPhilipse, I meant if it's supported in `SELECT REGEXP('abc', '([a-z]+)');` way.

No,it's a different way.

**Comment by maropu** [comment]
I get a bit confiused and Hive really supports REGEX as a function?
```
// REGEXP case in hive (3.1.1)
hive> SELECT 'abc' REGEXP '([a-z]+)';
OK
true

hive> select REGEXP('abc','([a-z]+)');
NoViableAltException(251@[])
	at org.apache.hadoop.hive.ql.parse.HiveParser_SelectClauseParser.selectClause(HiveParser_SelectClauseParser.java:962)
    ...
FAILED: ParseException line 1:7 cannot recognize input near 'REGEXP' '(' ''abc'' in select clause

// RLIKE in hive
hive> SELECT 'abc' RLIKE '([a-z]+)';
OK
true

hive> select RLIKE('abc','([a-z]+)');
NoViableAltException(265@[])
	at org.apache.hadoop.hive.ql.parse.HiveParser_SelectClauseParser.selectClause(HiveParser_SelectClauseParser.java:962)
	...
FAILED: ParseException line 1:7 cannot recognize input near 'RLIKE' '(' ''abc'' in select clause

// REGEXP in mysql
mysql> SELECT 'abc' REGEXP '([a-z]+)';
+-------------------------+
| 'abc' REGEXP '([a-z]+)' |
+-------------------------+
|                       1 |
+-------------------------+

mysql> select REGEXP('abc','([a-z]+)');
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'REGEXP('abc','([a-z]+)')' at line 1

// RLIKE in mysql
mysql> SELECT 'abc' RLIKE '([a-z]+)';
+------------------------+
| 'abc' RLIKE '([a-z]+)' |
+------------------------+
|                      1 |
+------------------------+

mysql> select RLIKE('abc','([a-z]+)');
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'RLIKE('abc','([a-z]+)')' at line 1
```

**Comment by GuoPhilipse** [comment]
I tried in hive 1.1, and it works. seems hive is incompatible for old version
`hive>  select RLIKE('abc','([a-z]+)');`
`true`

`hive> select REGEXP('abc','([a-z]+)');`
`true`

**Comment by maropu** [comment]
Yea, I checked the doc @HyukjinKwon put above, and it seems the current hive only supports `REGEXP` only in a SQL syntax.

File: docs/sql-ref-syntax-qry-select-like.md [hunk 2]
```diff
@@ -26,7 +26,7 @@ A LIKE predicate is used to search for a specific pattern.
 ### Syntax
 
 ```sql
-[ NOT ] { LIKE search_pattern [ ESCAPE esc_char ] | RLIKE regex_pattern }
+[ NOT ] { LIKE search_pattern [ ESCAPE esc_char ] | RLIKE regex_pattern | REGEXP regex_pattern}
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-like.md:29]: `[ RLIKE | REGEXP ] regex_pattern` cc: @huaxingao

File: docs/sql-migration-guide.md [hunk 3]
```diff
@@ -30,6 +30,8 @@ license: |
   
   - In Spark 3.1, `from_unixtime`, `unix_timestamp`,`to_unix_timestamp`, `to_timestamp` and `to_date` will fail if the specified datetime pattern is invalid. In Spark 3.0 or earlier, they result `NULL`.
 
+  - In Spark 3.1, we can use regexp function, it is the alias of rlike funtion, which functions the same with rike funtion.
```
Comment (maropu) [inline docs/sql-migration-guide.md:33]: We don't need this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 4]
```diff
@@ -359,6 +359,7 @@ object FunctionRegistry {
     expression[StringReplace]("replace"),
     expression[Overlay]("overlay"),
     expression[RLike]("rlike"),
+    expression[RLike]("regexp", true),
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:362]: Need this? It seems hive and mysql doesn't support this though.

File: docs/sql-ref-syntax-qry-select-like.md (hunk 2 above)
Comment (maropu) [reply docs/sql-ref-syntax-qry-select-like.md:29]: I think its okay for this PR to just update this doc.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 4 above)
Comment (GuoPhilipse) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:362]: yes, let me will remove the function usage

**Comment by HyukjinKwon** [comment]
ok to test

File: docs/sql-ref-syntax-qry-select-like.md (hunk 2 above)
Comment (GuoPhilipse) [reply docs/sql-ref-syntax-qry-select-like.md]: have updated and added examples for this usage.

**Review by HyukjinKwon — APPROVED** [review]
Looks okay but let me leave it to @maropu

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
Thanks! Merged to master/3.0.

File: docs/sql-ref-syntax-qry-select-like.md (hunk 2 above)
Comment (maropu) [reply docs/sql-ref-syntax-qry-select-like.md]: Btw, this reminded me that there are still missing keywords in the SQL docs: https://issues.apache.org/jira/browse/SPARK-31753
Its very helpful if you take this over.

Comment (GuoPhilipse) [reply docs/sql-ref-syntax-qry-select-like.md]: I am glad to take this over,will work on it later :)

Comment (maropu) [reply docs/sql-ref-syntax-qry-select-like.md]: Thanks! That's very helpful.

**Pull request closed without merging**